drakrun/drakrun/lib/postprocessing/capa_plugin/capa_processor.py

```diff
@@ -0,0 +1,428 @@
+    name = rule.name.split("/")[0]
+    mbc = rule.meta.get("mbc", None)
+    attck = rule.meta.get("att&ck", None)
+    occurrences = [format_capa_address(address=address) for address, _ in addresses]
+    occurrences = [
+        dict(t) for t in {tuple(d.items()) for d in occurrences}
+    ]  # remove duplicate addresses
+
+    ttp_block = dict()
+    ttp_block.update({"name": name})
+    ttp_block.update({"mbc": mbc} if mbc else {})
+    ttp_block.update({"att&ck": attck} if attck else {})
+    ttp_block.update({"occurrences": occurrences})
+
+    return ttp_block
+
+
+def construct_ttp_blocks(
+    rules: capa.rules.RuleSet,
+    capabilities_per_file: List[Tuple[Path, capa.engine.MatchResults, Any]],
+    filter_function=lambda rule: rule,
+) -> Iterator[Dict[str, Any]]:
+
+    # construct a ttp block for each extracted capability
+    for _, capabilities in capabilities_per_file:
+        for name, addresses in capabilities.items():
+            if filter_function(rules[name]):
+                yield construct_ttp_block(rules[name], addresses)
+
+
+def capa_analysis(analysis_dir: Path) -> None:
+    # check and prepare the rules folder
+    if not check_rules_directory_exist(capa_rules_dir):
+        # in case of a missing/empty rules folder, clone the official capa rules
+        logger.exception("capa rules directory is empty or non-existant")
+
+    # get rules and filter them
+    rules = get_rules([capa_rules_dir])
+    rules = filter_rules(
+        rules, filter_function=lambda rule: rule.meta.get("att&ck", None)
+    )  # select only rules with an att&ck entry specified
+
+    # get malware-related pids if requested by configuration
+    malware_pids = None
+    if analyze_malware_pids_only:
+        malware_pids = list(
+            get_malware_processes(
+                metadata_path=analysis_dir / "metadata.json",
+                inject_path=analysis_dir / "inject.log",
+                pstree_path=analysis_dir / "process_tree.json",
+            )
+        )
+
+    # make sure either static or dynamic capability extraction is on
+    assert perform_dynamic_analysis or perform_static_analysis
+
+    # extract capabilities from the Drakvuf report
+    if perform_dynamic_analysis:
+        dynamic_capabilities = dynamic_capa_analysis(
+            analysis_dir, rules, malware_pids=malware_pids
+        )
+
+        # write the extracted TTPs to the analysis dir
+        with (analysis_dir / "ttps.json").open("wb") as f:
+            for ttp in construct_ttp_blocks(
+                rules,
+                [dynamic_capabilities],
+                filter_function=lambda rule: rule.meta.get("att&ck", None),
+            ):
+                f.write(orjson.dumps(ttp))
+                f.write(b"\n")
+
+    # extract capabilities from the memory dumps
+    if perform_static_analysis:
+        static_capabilities_per_file = static_memory_dumps_capa_analysis(
+            analysis_dir, rules, malware_pids=malware_pids
```

Comment:
malware_pids can be None here, but in `static_memory_dumps_capa_analysis` the type is List[int] (and it doesn't handle empty list of arguments well).

Not a problem though, since malware_pids should be changed to non-nullable (as per other suggestions)